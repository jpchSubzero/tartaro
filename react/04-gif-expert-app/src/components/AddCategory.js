//Snippet -> rafc
import React, { useState } from 'react';

import { PropTypes } from 'prop-types';


//UpperCamelCase (más conocido como PascalCase) para saber que es un componente
// export const AddCategory = () => {

// Recibir función desde el padre y con destructuración asignarle el nombre
export const AddCategory = ({ setCategories }) => {

    // Tarea 2
    //     Recibir la función de agregar categoría
    //     Agregar una categoría al presionar enter

    // Tarea 3
    // Agregar propTypes para requerir la función desde el padre
    // Validar el error cuando no se envía

    //Quitamos el valor inicial, esto dará un error nuevo en consola porque inputValue va a ser undefined y se necesita un valor en <input ... value={inputValue} ... />>
    // const [inputValue, setInputValue] = useState('Valor inicial');

    // Para mostrar el error de valor undefined
    // const [inputValue, setInputValue] = useState();

    // Para corregir el error de undefined que es lo mismo que se hacia en angular, dar un valor inicial
    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        // Para visualizar el value desde el target del evento
        // console.log(e.target.value);
        setInputValue(e.target.value);
    }

    // Para manejar presionar enter
    const handleSubmit = (e) => {
        // Para prevenir el comportamiento por defecto del evento
        e.preventDefault();

        console.log('Submit');

        //Invocar la función recibida desde el padre y utilizarla a través del callback porque no tenemos acceso a categories, se puede recibir también las categories pero gracias al callback podemos pedirlas
        // setCategories(categs => [...categs, inputValue]);

        // Agregar validaciones porque acepta vacíos
        if (inputValue.trim()) {
            // Quitar agregar al final para que se vayan agregando al inicio
            // setCategories(categs => [...categs, inputValue]);            

            // Quitar agregar al final para que se vayan agregando al inicio
            setCategories(categs => [inputValue, ...categs]);            

            setInputValue('');
        }
    }

    return (
        // Quitamos <> porque forms ya engloba todos los demás tags
        // <>

        // Agregamos form para poder utilizar el evento submit
        <form onSubmit={ handleSubmit }>
            {/* Para verificar el componente, como AddCategory works!!! en angular */}
            {/* <h2>AddCategory</h2> */}

            {/* Agregar para mostrar el valor */}
            <h1>{ inputValue }</h1>

            {/* Agregar CSS si se quiere */}
            <input 
                type="text" 
                // Produce un error ya que quieres editar un valor inmutable (inputValue)
                value={inputValue}
                // Quita el error pero aún no es editable, debemos crear un manejador para ese evento y allí actualizarlo con el useState (setInputValue)
                // onChange={ (e) => console.log(e) }

                // El evento llama a handleInputChange para actualizar, como es solo una instrucción y recibe e de evento no es necesaria la expresión lambda
                onChange={ handleInputChange }

                // Agregar placeholder después de quitar el valor por defecto
                placeholder="Ingrese una categoría"
            />
        </form>
        // </>
    )
}

AddCategory.propTypes = {
    setCategories:PropTypes.func.isRequired
}