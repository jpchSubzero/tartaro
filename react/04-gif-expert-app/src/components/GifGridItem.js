import React from 'react';

// export const GifGridItem = (props) => {

//Para no usar en el componente props.image enviamos destructurado el objeto
export const GifGridItem = ({id, title, url}) => {

    console.log(title);

    return (
        // <div className="card animate__animated animate__bounce">

        // Agregar animate.css
        <div className="card animate__animated animate__fadeIn">
            {/* Para revisar si lleg√≥ la imagen */}
            {/* { title } */}
            <img src={url} alt={title} />
            <p>{ title }</p>
        </div>
    )
}
