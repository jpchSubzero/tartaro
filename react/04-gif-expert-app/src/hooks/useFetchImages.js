import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

//Snippet => rafc y luego borrar el import porque no se necesita react para JSX y quitar el return
// export const useFetchImages = () => {

//Recibir la categoría
export const useFetchImages = (category) => {
        //Se usa state porque el hook no sabe para que va a trabajar y por convención se usa state en lugar de otro nombre de variable
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    //Agregar el setTimeout para mostrar como cambia el mensaje cargado
    // setTimeout( () =>  {
    //     setState({
    //         data: [1,2,3,4,5,6],
    //         loading: false    
    //     });
    // }, 3000);

    // Comentar el setTimeout
    //Agergar el useEffect para controlar que solo se ejecute cuando queremos, en esta ocasión cuando cambia la categoría
    //Los useEffect no pueden ser async, siempre esperan un valor síncrono
    useEffect(() => {

            getGifs(category).then(resolve => {

                //Agregarmos el setTimeout solo para ver como se ejecuta
                //Comentar al final y solo hacer el useState
                // setTimeout( () =>  {
                //     console.log(resolve);
                //     setState({
                //         data: resolve,
                //         loading: false
                //     });
                // }, 3000);

                setState({
                    data: resolve,
                    loading: false
                });

            });
    }, [category]);


    return state;
}
