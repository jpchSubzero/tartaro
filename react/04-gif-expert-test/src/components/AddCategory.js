//Snippet -> rafc
import React, { useState } from 'react';

import { PropTypes } from 'prop-types';

export const AddCategory = ({ setCategories }) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e) => {
        setInputValue(e.target.value);

        //Agregar para prueba unitaria
        console.log('Se llamó el método handleInputChange');
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        //Agregar para prueba unitaria
        console.log('Se llamó el método handleSubmit', inputValue);

        if (inputValue.trim().length > 2) {
            setCategories(categs => [inputValue, ...categs]);   
            // setCategories(2412341234);            
            setInputValue('');
        }
    }

    return (
        <form onSubmit={ handleSubmit }>
            <h1>{ inputValue }</h1>

{/* Agregamos el <p> por pruebas unitarias */}
<p>{inputValue }</p>

            <input 
                type="text" 
                value={inputValue}
                onChange={ handleInputChange }
                placeholder="Ingrese una categoría"
            />
        </form>
        // </>
    )
}

AddCategory.propTypes = {
    setCategories:PropTypes.func.isRequired
}