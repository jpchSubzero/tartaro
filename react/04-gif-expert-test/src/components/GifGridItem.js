import React from 'react';

//Agregar propTypes para que envien title y url obligatoriamente
//snapshot -> impt
import PropTypes from 'prop-types';

export const GifGridItem = ({title, url}) => {

    console.log(title);

    return (
        <div className="card animate__animated animate__fadeIn">
            <img src={url} alt={title} />
            <p>{ title }</p>
        </div>
    )
}

//Agregar propTypes para que envien title y url obligatoriamente
GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
}

// Tarea 1
// Enzyme
// Enzume to json
// Test para GifGridItem
//     shallow
//     wrapper
//     wrapper .toMatchSnapshot()

// Tarea 2
// Agregar propTypes para que envien title y url obligatoriamente
// Enviar los parametros en la prueba
// Actualizar el snapshot

// Tarea 3
// Validar que el componente tenga la clase animate__fadeIn