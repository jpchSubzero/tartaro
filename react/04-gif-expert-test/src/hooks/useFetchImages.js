import { useState, useEffect } from 'react';
import { getGifs } from '../helpers/getGifs';

// Se puede validar que al inicio exista data vacÃ­a y loading en true. TambiÃ©n que luego del effect exista data y loading en false.

export const useFetchImages = (category) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    useEffect(() => {
            getGifs(category).then(resolve => {
                setState({
                    data: resolve,
                    loading: false
                });

            });
    }, [category]);

    return state;
}
