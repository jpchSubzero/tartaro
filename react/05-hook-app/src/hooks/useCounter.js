// snippet -> rafc y quitar lo que no sirve
import { useState } from 'react';

// Creamos un hook que tiene un valor por defecto
export const useCounter = ( initialState = 10 ) => {
    const [state, setState] = useState(initialState);

    // const increment = () => {

    // Agregar un factor de suma
    const increment = (factor = 1) => {
        // setState(state + 1);

        // Cambiar el 1 por el factor
        setState(state + factor);
    };

    // const decrement = (factor = 1) => {

    // Agregar un factor de resta
    const decrement = (factor = 1) => {
        // setState(state - 1);

        // Cambiar el 1 por el factor
        setState(state - factor);
    };

    // Resoluci贸n de tarea. Agregar funci贸n de reset
    const reset = () => {
        setState(initialState);
    };
    
    // Hacemos accesibles las propiedades y funciones del hook
    return {
        state,
        increment,
        decrement,

        // Resoluci贸n de tarea. Devolver funci贸n de reset
        reset
    };
}
