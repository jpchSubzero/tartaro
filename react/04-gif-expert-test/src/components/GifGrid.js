import React from 'react';
import { useFetchImages } from '../hooks/useFetchImages';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({ category }) => {

    const { loading, data:images } = useFetchImages(category);

    return (
        <>
            <h3 className="animate__animated animate__fadeIn">{ category }</h3>
            { loading && <p className="animate__animated animate__flash">Cargando...</p> }

            <div className="card-grid">
            {
                    images.map((image) => {
                        if (!image.title) {
                            return '';
                        }
                        return (
                            <GifGridItem 
                                key={image.id}
                                {...image}
                            />
                        );                    
                    })
                }
            </div>
        </>
    )
}

// Agregar propTypes para obligar que se envie category
GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}

// Tarea 1
// Crear una prueba unitaria para evaluar el estado inicial del componente comparando con el snapshot