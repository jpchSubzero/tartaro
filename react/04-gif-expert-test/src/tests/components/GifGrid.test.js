//Importar para que funcione enzyme
import { shallow } from 'enzyme';

//Importar para poder utilizar JSX
import React from 'react';

import { GifGrid } from '../../components/GifGrid';

//Importamos el hook para poder implementar el mock
import { useFetchImages } from '../../hooks/useFetchImages';
jest.mock('../../hooks/useFetchImages');

describe('Pruebas con el archivo GifGrid', () => {

    // Como category se puso requerido le pasamos de una vez para que no de el error
    const category = 'saint seiya';

    test('Debe ser igual al snapshot en el arranque del componente', () => {
        // Como va a ser con el estado inicial, dejamos el loading en cargando y la data en vacío
        useFetchImages.mockReturnValue({
            data: [],
            loading: true            
        });

        // Al implementar el mock esta prueba va a fallar ya que piensa que hay data desde el hook pero aún no se ha mockeado
        // Para superar hay que agregar antes el mockeo de la data del hook
        const wrapper = shallow(<GifGrid category  = {category}/>);
        expect(wrapper).toMatchSnapshot();
    });

    // Este test es correcto pero depende de la data que enviamos, si cambia la data implica actualizar el snapshot, por lo cual es mejor validar de otra forma
    test('Debe mostrar items cuando se cargan imágenes con useFetchGifs y no mostrar loading. A través del snapshot', () => {
        // Creamos un mock para la data ya que, en este momento, no es de interés probar el hook
        // Como va a ser un estado de información ya cargada, llenamos la data con cualquier valor acorde y el loading en false
        const gifs = [
            {
                id: '1',
                title: 'Gif 1',
                url: 'http://image1'                
            },
            {
                id: '2',
                title: 'Gif 2',
                url: 'http://image2'
            },
            {
                id: '3',
                title: 'Gif 3',
                url: 'http://image3'
            }
        ];

        useFetchImages.mockReturnValue({
            data: gifs,
            loading: false            
        });

        const wrapper = shallow(<GifGrid category  = {category}/>);
        expect(wrapper).toMatchSnapshot();
    });


    test('Debe mostrar items cuando se cargan imágenes con useFetchGifs y no mostrar loading, a través de', () => {
        // Creamos un mock para la data ya que, en este momento, no es de interés probar el hook
        // Como va a ser un estado de información ya cargada, llenamos la data con cualquier valor acorde y el loading en false
        const gifs = [
            {
                id: '1',
                title: 'Gif 1',
                url: 'http://image1'                
            },
            {
                id: '2',
                title: 'Gif 2',
                url: 'http://image2'
            },
            {
                id: '3',
                title: 'Gif 3',
                url: 'http://image3'
            }
        ];

        const loading = false;
        useFetchImages.mockReturnValue({
            data: gifs,
            loading: loading            
        });

        const wrapper = shallow(<GifGrid category  = {category}/>);

        if (!loading) {
            // Validar que no se muestre el Cargando... si loading es false. Esto hacemos buscando el <p> ya que si loading es false no se presenta esa etiqueta
            expect(wrapper.find('p').exists()).toBe(false);

            // Validar que la etiqueta GifGridItem tenga el mismo número de elementos que la data que enviamos
            expect(wrapper.find('GifGridItem').length).toBe(gifs.length);            
        } else {
            // Validar que se muestre el Cargando... si loading es true. Esto hacemos buscando el <p> ya que si loading es true se presenta esa etiqueta
            expect(wrapper.find('p').exists()).toBe(true);

            // Validar que la etiqueta GifGridItem no exista cuando se están cargando los elementos
            expect(wrapper.find('GifGridItem').exists()).toBe(false);            

        }
    });    
    
})
