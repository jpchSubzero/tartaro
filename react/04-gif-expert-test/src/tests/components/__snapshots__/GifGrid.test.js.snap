// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con el archivo GifGrid Debe mostrar items cuando se cargan imágenes con useFetchGifs y no mostrar loading. A través del snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    saint seiya
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="1"
      key="1"
      title="Gif 1"
      url="http://image1"
    />
    <GifGridItem
      id="2"
      key="2"
      title="Gif 2"
      url="http://image2"
    />
    <GifGridItem
      id="3"
      key="3"
      title="Gif 3"
      url="http://image3"
    />
  </div>
</Fragment>
`;

exports[`Pruebas con el archivo GifGrid Debe ser igual al snapshot en el arranque del componente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    saint seiya
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
